src/js/gutenberg-plugins/inserter-media-category.js

```diff
@@ -30,38 +30,65 @@ waitFor( isInserterOpened ).then( () =>
 	)
 );
 
+/**
+ * A slightly modified debounced function to add delay
+ * to an already debounced function.
+ *
+ * {@link https://github.com/10up/classifai/issues/561}
+ * {@link https://github.com/10up/classifai/pull/535}
+ *
+ * @param {Function} func    The function to be debounced.
+ * @param {number}   timeout The delay in milliseconds.
+ * @return {Function} The debounced function.
+ */
+const debounce = ( func, timeout = 250 ) => {
+	let timer;
+
+	return ( ...args ) => {
+		clearTimeout( timer );
+
+		return new Promise( ( resolve ) => {
+			timer = setTimeout( () => {
+				resolve( func.apply( this, args ) );
+			}, timeout );
+		} );
+	};
+};
+
+const imageFetcher = async ( { search = '' } ) => {
+	if ( ! search ) {
+		return [];
+	}
+
+	const images = await apiFetch( {
+		path: addQueryArgs( classifaiDalleData.endpoint, {
+			prompt: search,
+			format: 'b64_json',
+		} ),
+		method: 'GET',
+	} )
+		.then( ( response ) =>
+			response.map( ( item ) => ( {
+				title: search,
+				url: `data:image/png;base64,${ item.url }`,
+				previewUrl: `data:image/png;base64,${ item.url }`,
+				id: undefined,
+				alt: search,
+				caption: classifaiDalleData.caption,
+			} ) )
+		)
+		.catch( () => [] );
+
+	return images;
+};
+
 const registerGenerateImageMediaCategory = () => ( {
 	name: 'classifai-generate-image',
 	labels: {
 		name: classifaiDalleData.tabText,
 		search_items: __( 'Enter a prompt', 'classifai' ),
 	},
 	mediaType: 'image',
-	fetch: async ( { search = '' } ) => {
-		if ( ! search ) {
-			return [];
-		}
-
-		const images = await apiFetch( {
-			path: addQueryArgs( classifaiDalleData.endpoint, {
-				prompt: search,
-				format: 'b64_json',
-			} ),
-			method: 'GET',
-		} )
-			.then( ( response ) =>
-				response.map( ( item ) => ( {
-					title: search,
-					url: `data:image/png;base64,${ item.url }`,
-					previewUrl: `data:image/png;base64,${ item.url }`,
-					id: undefined,
-					alt: search,
-					caption: classifaiDalleData.caption,
-				} ) )
-			)
-			.catch( () => [] );
-
-		return images;
-	},
+	fetch: debounce( imageFetcher, 700 ),
```

Comment:
I think we need to increase this delay. In my testing, I was still seeing multiple ajax requests at this point. If I increase this up to 2000 or 3000, I still occasionally saw two ajax requests but often only one and never more than two

Reply:
@dkotter I performed a programmatic test by simulating human input every "x" seconds.

## Case 1: Debounce delay: 1000ms, typing interval: 900. Result - No ajax calls


https://github.com/10up/classifai/assets/17757960/b3e21228-bf92-40d2-967c-8aa17e4e4906


## Case 2: Debounce delay: 1000ms, typing interval: 1100. Result - Ajax calls for every input change

https://github.com/10up/classifai/assets/17757960/b1fd2ec9-a297-416c-b5f1-73c22099ff2b

Just wanted to confirm if we should be increasing the debounce delay between 2000 - 3000ms?

Reply:
So here's what I get right now:

https://github.com/10up/classifai/assets/916738/8aa12b8d-42fc-4639-a427-75fbc889cbac

I enter a fairly simple image prompt with a fairly normal typing speed and I end up getting 7 requests before I actually finish (which at 2 cents per image and generating 3 images per request, that ends up costing 42 cents).

If I change the debounce all the way up to 3000, I get only a single request, though it takes a bit for that request to fire off:

https://github.com/10up/classifai/assets/916738/fb513f7d-c21d-4d34-a7b3-59fa0a4f701b

So unless I'm testing this wrong, I think we need to increase that debounce value somewhere between 2000 and 3000 to be more sure less requests will get sent.

Reply:
Got it, it's the pauses a user may take after typing few words that will trigger AJAX requests. I've increased the delay to 2500ms.